File: src/sbml/test/TestParameterRule.c
/**
 * \file    TestParameterRule.c
 * \brief   ParameterRule unit tests
 * \author  Ben Bornstein
 *
 * $Id$
 * $Source$
 */
/* Copyright 2002 California Institute of Technology and
 * Japan Science and Technology Corporation.
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2.1 of the License, or
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY, WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  The software and
 * documentation provided hereunder is on an "as is" basis, and the
 * California Institute of Technology and Japan Science and Technology
 * Corporation have no obligations to provide maintenance, support,
 * updates, enhancements or modifications.  In no event shall the
 * California Institute of Technology or the Japan Science and Technology
 * Corporation be liable to any party for direct, indirect, special,
 * incidental or consequential damages, including lost profits, arising
 * out of the use of this software and its documentation, even if the
 * California Institute of Technology and/or Japan Science and Technology
 * Corporation have been advised of the possibility of such damage.  See
 * the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 *
 * The original code contained here was initially developed by:
 *
 *     Ben Bornstein
 *     The Systems Biology Markup Language Development Group
 *     ERATO Kitano Symbiotic Systems Project
 *     Control and Dynamical Systems, MC 107-81
 *     California Institute of Technology
 *     Pasadena, CA, 91125, USA
 *
 *     http://www.cds.caltech.edu/erato
 *     mailto:sbml-team@caltech.edu
 *
 * Contributor(s):
 */


#include "common/common.h"

#include "SBase.h"
#include "Rule.h"

#include <check.h>


static Rule_t *PR;


void
ParameterRuleTest_setup (void)
{
  PR = Rule_createAssignment();
  Rule_setL1TypeCode(PR, SBML_PARAMETER_RULE);

  if (PR == NULL)
  {
    fail("ParameterRule_create() returned a NULL pointer.");
  }
}


void
ParameterRuleTest_teardown (void)
{
  Rule_free(PR);
}


START_TEST (test_ParameterRule_create)
{
  fail_unless( SBase_getTypeCode((SBase_t *) PR) ==
               SBML_ASSIGNMENT_RULE );
  fail_unless( Rule_getL1TypeCode((Rule_t *) PR) ==
               SBML_PARAMETER_RULE );
  fail_unless( SBase_getNotes     ((SBase_t *) PR) == NULL );
  fail_unless( SBase_getAnnotation((SBase_t *) PR) == NULL );

  fail_unless( Rule_getFormula((Rule_t *) PR) == NULL );

  fail_unless( Rule_getUnits(PR) == NULL );
  fail_unless( Rule_getVariable (PR) == NULL );

  fail_unless( Rule_getType( PR) ==  RULE_TYPE_SCALAR );

  fail_unless( !Rule_isSetVariable (PR) );
  fail_unless( !Rule_isSetUnits(PR) );
}
END_TEST


START_TEST (test_ParameterRule_createWith)
{
  Rule_t *pr;


  pr = Rule_createRateWithVariableAndFormula("c", "v + 1");
  Rule_setL1TypeCode(pr, SBML_PARAMETER_RULE);

  fail_unless( SBase_getTypeCode((SBase_t *) pr) ==
               SBML_RATE_RULE );
  fail_unless( Rule_getL1TypeCode((Rule_t *) pr) ==
               SBML_PARAMETER_RULE );
  fail_unless( SBase_getNotes     ((SBase_t *) pr) == NULL );
  fail_unless( SBase_getAnnotation((SBase_t *) pr) == NULL );

  fail_unless( Rule_getUnits(pr) == NULL );

  fail_unless( !strcmp(Rule_getFormula(pr), "v + 1") );
  fail_unless( !strcmp(Rule_getVariable(pr), "c") );

  fail_unless( Rule_getType( pr) ==  RULE_TYPE_RATE );

  fail_unless( Rule_isSetVariable(pr) );
  fail_unless( !Rule_isSetUnits(pr) );

  Rule_free(pr);
}
END_TEST


START_TEST (test_ParameterRule_free_NULL)
{
  Rule_free(NULL);
}
END_TEST


START_TEST (test_ParameterRule_setName)
{
  char *name = "cell";
  const char *c;


  Rule_setVariable(PR, name);

  fail_unless( !strcmp(Rule_getVariable(PR), name));
  fail_unless( Rule_isSetVariable(PR) );

  if (Rule_getVariable(PR) == name)
  {
    fail( "ParameterRule_setName(...) did not make a copy of string." );
          
  }

  /* Reflexive case (pathological) */
  c = Rule_getVariable(PR);
  Rule_setVariable(PR, c);
  fail_unless( !strcmp(Rule_getVariable(PR), name),
               NULL );

  Rule_setVariable(PR, NULL);
  fail_unless( !Rule_isSetVariable(PR) );

  if (Rule_getVariable(PR) != NULL)
  {
    fail( "Rule_setVariable(PR, NULL)"
          " did not clear string." );
  }
}
END_TEST


START_TEST (test_ParameterRule_setUnits)
{
  char *units = "cell";


  Rule_setUnits(PR, units);

  fail_unless( !strcmp(Rule_getUnits(PR), units)    );
  fail_unless( Rule_isSetUnits(PR) );

  if (Rule_getUnits(PR) == units)
  {
    fail( "Rule_setUnits(...) did not make a copy of string." );
  }

  /* Reflexive case (pathological) */
  Rule_setUnits(PR, Rule_getUnits(PR));
  fail_unless( !strcmp(Rule_getUnits(PR), units) );

  Rule_setUnits(PR, NULL);
  fail_unless( !Rule_isSetUnits(PR) );

  if (Rule_getUnits(PR) != NULL)
  {
    fail( "Rule_setUnits(PR, NULL) did not clear string." );
  }
}
END_TEST


Suite *
create_suite_ParameterRule (void)
{
  Suite *suite = suite_create("ParameterRule");
  TCase *tcase = tcase_create("ParameterRule");


  tcase_add_checked_fixture( tcase,
                             ParameterRuleTest_setup,
                             ParameterRuleTest_teardown );

  tcase_add_test( tcase, test_ParameterRule_create     );
  tcase_add_test( tcase, test_ParameterRule_createWith );
  tcase_add_test( tcase, test_ParameterRule_free_NULL  );
  tcase_add_test( tcase, test_ParameterRule_setName    );
  tcase_add_test( tcase, test_ParameterRule_setUnits   );

  suite_add_tcase(suite, tcase);

  return suite;
}

File: src/bindings/csharp/test/sbml/TestParameterRule.cs
/// 
///  @file    TestParameterRule.cs
///  @brief   ParameterRule unit tests
///  @author  Frank Bergmann (Csharp conversion)
///  @author  Akiya Jouraku (Csharp conversion)
///  @author  Ben Bornstein 
/// 
///  $Id:$
///  $HeadURL:$
/// 
///  This test file was converted from src/sbml/test/TestParameterRule.c
///  with the help of conversion sciprt (ctest_converter.pl).
/// 
/// <!---------------------------------------------------------------------------
///  This file is part of libSBML.  Please visit http://sbml.org for more
///  information about SBML, and the latest version of libSBML.
/// 
///  Copyright 2005-2009 California Institute of Technology.
///  Copyright 2002-2005 California Institute of Technology and
///                      Japan Science and Technology Corporation.
///  
///  This library is free software; you can redistribute it and/or modify it
///  under the terms of the GNU Lesser General Public License as published by
///  the Free Software Foundation.  A copy of the license agreement is provided
///  in the file named "LICENSE.txt" included with this software distribution
///  and also available online as http://sbml.org/software/libsbml/license.html
/// --------------------------------------------------------------------------->*/


namespace LibSBMLCSTest {

  using libsbml;

  using  System.IO;

  public class TestParameterRule {
    public class AssertionError : System.Exception 
    {
      public AssertionError() : base()
      {
        
      }
    }


    static void assertTrue(bool condition)
    {
      if (condition == true)
      {
        return;
      }
      throw new AssertionError();
    }

    static void assertEquals(object a, object b)
    {
      if ( (a == null) && (b == null) )
      {
        return;
      }
      else if (a.Equals(b))
      {
        return;
      }
  
      throw new AssertionError();
    }

    static void assertNotEquals(object a, object b)
    {
      if ( (a == null) && (b == null) )
      {
        throw new AssertionError();
      }
      else if (a.Equals(b))
      {
        throw new AssertionError();
      }
    }

    static void assertEquals(bool a, bool b)
    {
      if ( a == b )
      {
        return;
      }
      throw new AssertionError();
    }

    static void assertNotEquals(bool a, bool b)
    {
      if ( a != b )
      {
        return;
      }
      throw new AssertionError();
    }

    static void assertEquals(int a, int b)
    {
      if ( a == b )
      {
        return;
      }
      throw new AssertionError();
    }

    static void assertNotEquals(int a, int b)
    {
      if ( a != b )
      {
        return;
      }
      throw new AssertionError();
    }

    private Rule PR;

    public void setUp()
    {
      PR = new  AssignmentRule();
      PR.setL1TypeCode(libsbml.SBML_PARAMETER_RULE);
      if (PR == null);
      {
      }
    }

    public void tearDown()
    {
      PR = null;
    }

    public void test_ParameterRule_create()
    {
      assertTrue( PR.getTypeCode() == libsbml.SBML_ASSIGNMENT_RULE );
      assertTrue( PR.getL1TypeCode() == libsbml.SBML_PARAMETER_RULE );
      assertTrue( PR.getNotes() == null );
      assertTrue( PR.getAnnotation() == null );
      assertTrue( PR.getFormula() == "" );
      assertTrue( PR.getUnits() == "" );
      assertTrue( PR.getVariable() == "" );
      assertTrue( PR.getType() == libsbml.RULE_TYPE_SCALAR );
      assertEquals( false, PR.isSetVariable() );
      assertEquals( false, PR.isSetUnits() );
    }

    public void test_ParameterRule_createWith()
    {
      Rule pr;
      pr = new  RateRule("c", "v + 1");
      pr.setL1TypeCode(libsbml.SBML_PARAMETER_RULE);
      assertTrue( pr.getTypeCode() == libsbml.SBML_RATE_RULE );
      assertTrue( pr.getL1TypeCode() == libsbml.SBML_PARAMETER_RULE );
      assertTrue( pr.getNotes() == null );
      assertTrue( pr.getAnnotation() == null );
      assertTrue( pr.getUnits() == "" );
      assertTrue((  "v + 1" == pr.getFormula() ));
      assertTrue((  "c" == pr.getVariable() ));
      assertTrue( pr.getType() == libsbml.RULE_TYPE_RATE );
      assertEquals( true, pr.isSetVariable() );
      assertEquals( false, pr.isSetUnits() );
      pr = null;
    }

    public void test_ParameterRule_free_NULL()
    {
    }

    public void test_ParameterRule_setName()
    {
      string name =  "cell";;
      string c;
      PR.setVariable(name);
      assertTrue(( name == PR.getVariable() ));
      assertEquals( true, PR.isSetVariable() );
      if (PR.getVariable() == name);
      {
      }
      c = PR.getVariable();
      PR.setVariable(c);
      assertTrue(( name == PR.getVariable() ));
      PR.setVariable("");
      assertEquals( false, PR.isSetVariable() );
      if (PR.getVariable() != null);
      {
      }
    }

    public void test_ParameterRule_setUnits()
    {
      string units =  "cell";;
      PR.setUnits(units);
      assertTrue(( units == PR.getUnits() ));
      assertEquals( true, PR.isSetUnits() );
      if (PR.getUnits() == units);
      {
      }
      PR.setUnits(PR.getUnits());
      assertTrue(( units == PR.getUnits() ));
      PR.setUnits("");
      assertEquals( false, PR.isSetUnits() );
      if (PR.getUnits() != null);
      {
      }
    }

  }
}
